src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -301,14 +301,26 @@ internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, Diagnos
 
                     stateBuilder[i] = generatorState.WithResults(ParseAdditionalSources(state.Generators[i], sources, cancellationToken), generatorDiagnostics, generatorRunStateTable.ExecutedSteps, generatorRunStateTable.OutputSteps, hostOutputs, generatorTimer.Elapsed);
                 }
-                catch (UserFunctionException ufe)
+                catch (UserFunctionException ufe) when (handleGeneratorException(compilation, MessageProvider, state.Generators[i], ufe.InnerException, isInit: false))
                 {
                     stateBuilder[i] = SetGeneratorException(compilation, MessageProvider, generatorState, state.Generators[i], ufe.InnerException, diagnosticsBag, cancellationToken, runTime: generatorTimer.Elapsed);
                 }
             }
 
             state = state.With(stateTable: driverStateBuilder.ToImmutable(), syntaxStore: syntaxStoreBuilder.ToImmutable(), generatorStates: stateBuilder.ToImmutableAndFree(), runTime: timer.Elapsed, parseOptionsChanged: false);
             return state;
+
+            static bool handleGeneratorException(Compilation compilation, CommonMessageProvider messageProvider, ISourceGenerator sourceGenerator, Exception e, bool isInit)
+            {
+                if (!compilation.CatchAnalyzerExceptions)
+                {
+                    Debug.Assert(false);
+                    Environment.FailFast(CreateGeneratorExceptionDiagnostic(messageProvider, sourceGenerator, e, isInit).ToString());
```

Comment:
I'm curious why we bother creating the diagnostic at all in these cases? The diagnostic is just supposed to be a user-friendly representation of the exception, but here we actively _want_ the exception info. Do we actually gain anything by wrapping it in the diagnostic?

Reply:
It has the advantage of including the source generator name (and maybe some other info could be added in the future, so feels like this could be worth keeping in sync, I think there is some context info for analyzers passed in for example).